File: 프로그래머스/레벨3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        Map<String, Integer> genreMap = new HashMap<>();
        Map<String, List<int[]>> songMap = new HashMap<>();

        for (int i = 0; i < genres.length; i++) {
            genreMap.put(genres[i], genreMap.getOrDefault(genres[i], 0) + plays[i]);
            if (!songMap.containsKey(genres[i])) {
                songMap.put(genres[i], new ArrayList<>());
            }
            songMap.get(genres[i]).add(new int[]{plays[i], i});
        }

        List<String> sortedGenres = new ArrayList<>(genreMap.keySet());
        sortedGenres.sort((a, b) -> genreMap.get(b) - genreMap.get(a));

        List<Integer> result = new ArrayList<>();
        for (String genre : sortedGenres) {
            List<int[]> sortedSongs = songMap.get(genre);
            sortedSongs.sort((a, b) -> b[0] - a[0] == 0 ? a[1] - b[1] : b[0] - a[0]);
            for (int i = 0; i < Math.min(2, sortedSongs.size()); i++) {
                result.add(sortedSongs.get(i)[1]);
            }
        }

        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}

File: 프로그래머스/레벨3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const genreMap = new Map();
    const songMap = new Map();

    genres.forEach((genre, i) => {
        genreMap.set(genre, (genreMap.get(genre) || 0) + plays[i]);
        if (!songMap.has(genre)) songMap.set(genre, []);
        songMap.get(genre).push([plays[i], i]);
    });

    const sortedGenres = Array.from(genreMap.keys()).sort((a, b) => genreMap.get(b) - genreMap.get(a));

    const result = [];
    sortedGenres.forEach((genre) => {
        const sortedSongs = songMap.get(genre).sort((a, b) => b[0] - a[0] || a[1] - b[1]);
        result.push(...sortedSongs.slice(0, 2).map(song => song[1]));
    });

    return result;
}